### PR TITLE
FOGL-1654: C++ North plugins have incorrect API

### PR DESCRIPTION
**Pull request by pintomax**
WIP

This implementation lets "plugin_init" to use only ConfigCategory parameter
There is an extra API call "plugin_extra_config" at the moment.
We are looking for other solutions, avoiding usage of "plugin_extra_config".

Implementation details

C++ North plugins have incorrect API

- With this change the “plugin_init” method takes as parameter a ConfigCategory object instead of a JSON string map configuration. 

- The "PI_Server" north plugin only has new method “plugin_extra_config”.
   NOTE: if the method “plugin_extra_config” doesn’t exist and it's called, an empty string is returned: 
   no exceptions.

  `string extraConfig = this->m_plugin->extra_config();`

- When this method is called it returns a JSON with additional categories to create/update.
  Those category items (created/updated and fetched) are passed back to plugin via “plugin_init” as 
  new items with this name format:
  CATEGORY_NAME . INTEM_NAME, i.e.
  OMF_TYPES.type-id

   `string typesId = configData->getValue("OMF_TYPES.type-id");`

**_Additions:_**
get_plugin_info utility updated with “plugin_extra_config”:

plugins/utils/get_plugin_info plugins/north/PI_Server/libPI_Server.so.1 plugin_extra_config | jq

```
{
  "name": "Additional configuration",
  "description": "Additional configuration categories to pass to plugin_init",
  "categories": {
    "OMF_TYPES": {
      "type-id": {
        "description": "Identify sensor and measurement types",
        "type": "integer",
        "default": "0001"
      }
    }
  }
}
```

### CODE REVIEW

**Review by AmandeepArora — CHANGES_REQUESTED** [review]

File: C/plugins/north/PI_Server/plugin.cpp [hunk 1]
```diff
@@ -183,15 +193,16 @@ PLUGIN_HANDLE plugin_init(map<string, string>&& configData)
...
+				  PLUGIN_NAME,
 				  url.c_str(),
 				  producerToken.c_str(),
 				  typesId.c_str());
 
 
-	// TODO: return a more useful data structure for pluin handle
-	string* handle = new string("Init done");
+	// Return plugin handle
+	string* handle = new string(string(PLUGIN_NAME) + " initialised");
```
[the remark this answers is not in the document]
Comment (pintomax) [reply C/plugins/north/PI_Server/plugin.cpp]: This PI_Server plugin, the former OMF one, was the first C++ north plugin ever written.
I'll follow your suggestions, thanks.

File: C/plugins/north/PI_Server/plugin.cpp [hunk 2]
```diff
@@ -49,17 +50,10 @@ using namespace std;
...
-				"\"type\": \"boolean\", \"default\": \"False\" }, " \
-			"\"filterRule\": { " \
-				"\"description\": \"JQ formatted filter to apply (applicable if applyFilter is True)\", " \
-				"\"type\": \"string\", \"default\": \".[]\" }"
+				"\"type\": \"string\", \"default\": \"int64\" } " 
 
-
-
-#define OMF_PLUGIN_DESC "\"plugin\": {\"description\": \"OMF North C Plugin\", \"type\": \"string\", \"default\": \"omf\"}"
+#define OMF_PLUGIN_DESC "\"plugin\": {\"description\": \"OMF North C Plugin\", " \
```
[the remark this answers is not in the document]
Comment (pintomax) [reply C/plugins/north/PI_Server/plugin.cpp]: Yes, working on it

File: C/plugins/north/http-north/plugin.cpp [hunk 3]
```diff
@@ -51,21 +51,14 @@ extern "C" {
...
+	0,				// Flags
+	PLUGIN_TYPE_NORTH,		// Type
+	"1.0.0",			// Interface version
+	PLUGIN_DEFAULT_CONFIG_INFO	// Configuration
 };
 
-static const map<const string, const string> plugin_configuration = {
-					{
-						"PLUGIN",
-						string(PLUGIN_DEFAULT_CONFIG)
```
[the remark this answers is not in the document]
Comment (pintomax) [reply C/plugins/north/http-north/plugin.cpp]: Fixed

File: C/plugins/north/ThingSpeak/plugin.cpp [hunk 4]
```diff
@@ -75,58 +75,41 @@ PLUGIN_INFORMATION *plugin_info()
 	return &info;
 }
 
-static const map<const string, const string> plugin_configuration = {
-					{
-						"PLUGIN",
-						string(PLUGIN_DEFAULT_CONFIG)
```
[the remark this answers is not in the document]
Comment (pintomax) [reply C/plugins/north/ThingSpeak/plugin.cpp]: Fixed

File: C/plugins/north/PI_Server/plugin.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (pintomax) [reply C/plugins/north/PI_Server/plugin.cpp]: Fixed

**Review by AmandeepArora — APPROVED** [review]

File: C/tasks/north/sending_process/include/north_plugin.h [hunk 5]
```diff
@@ -32,17 +33,19 @@ class NorthPlugin : public Plugin {
...
 		uint32_t		send(const std::vector<Reading* >& readings) const;
-		PLUGIN_HANDLE		init(const std::map<std::string, std::string>& config);
+		PLUGIN_HANDLE		init(const ConfigCategory& config);
 
 	private:
 		// Function pointers
 		void			(*pluginShutdownPtr)(const PLUGIN_HANDLE);
-		std::map<const std::string, const std::string>&	(*pluginGetConfig)();
+		PLUGIN_INFORMATION*	(*pluginInfo)();
+		std::string&		(*pluginExtraConfig)();
```
Comment (MarkRiddoch) [inline C/tasks/north/sending_process/include/north_plugin.h]: Not happy with this extension to the API

**Review by MarkRiddoch — CHANGES_REQUESTED** [review]

**Comment by MarkRiddoch** [comment]
Having discussed this with Stefano it appears that it is feasible to make the handling of the increment automatic and not have to confuse the user with this parameter. All that remains is to define the way for the plugin to persist data between executions without using the configuration manager. The Python codes gets a handle to the storage layer, via an incorrect use of the plugin handle. Although not right it might signify that creating a legitimate mechanism for the plugin to get an handle on the storage layer is the best solution.

**Comment by pintomax** [comment]
Closing PR. Another will be created for North C++ API  fixes

**Pull request closed without merging**